### ISSUE TITLE
Is the Next.js documentation misleading?

### ISSUE THREAD

**Issue by nbouvrette** [PostHog/posthog#14307]
## In what situation are you experiencing subpar performance?

I followed the [Next.js documentation](https://posthog.com/docs/integrate/third-party/next-js) to integrate PostHog but noticed that my base application package size included the entire PostHog dependency that adds about **30kb** of gzipped data.

Is there a reason to import the PostHog package directly? Because I used dynamic imports instead and it seems to work fine:

```ts
import { useRouter } from 'next/router';
import { useEffect } from 'react';
import posthog from 'posthog-js';

// Dynamically imports PostHog to track page views
const handleRouteChange = () => {
  void import('posthog-js').then((module) => {
    const posthog = module.default

    posthog.init('<ph_project_api_key>', {
      api_host: '<ph_instance_address>',
      loaded: (posthog) => {
        if (process.env.NODE_ENV === 'development') {
          posthog.opt_out_capturing()
        } else {
          posthog.capture('$pageview')
        }
      },
    })
  })
}

function MyApp({ Component, pageProps }) {
  const router = useRouter();

  useEffect(() => {
    router.events.on('routeChangeComplete', handleRouteChange);

    return () => {
      router.events.off('routeChangeComplete', handleRouteChange);
    };
  }, []);

  return <Component {...pageProps} />;
}

export default MyApp;
```

## How to reproduce

Just follow the official doc: https://posthog.com/docs/integrate/third-party/next-js

Install and run [@next/bundle-analyzer](https://www.npmjs.com/package/@next/bundle-analyzer) to view the PostHog bundle size.

## Environment

- [ ] PostHog Cloud
- [ ] self-hosted PostHog, version/commit: _please provide_

**Comment by benjackwhite** [comment]
I'm not sure about "misleading". Do we suggest something somewhere that you felt mislead by? Perhaps that is a mistake in our docs 🤔 

Given that typically PostHog is one of the first things to be fired when the application loads (think initial `pageview`, `identifying` a user, loading feature flags) it is typically within the users interest to have the JS pre-loaded.
Either you load the bundled JS at the same time as the rest of your code, allowing PostHog to be available immediately or you load it later. It's up to you to decide if the extra complexity of dynamically loading the JS is worth it.

Not directly related but just FYI the example you posted will initialize PostHog multiple times which is almost definitely _not_ what you wan to do, but generally the idea of "promisifying" PostHog is completely fine. Something like:

```tsx
import { useRouter } from "next/router";
import { useEffect } from "react";

let posthogInitialised = false;

const asyncPosthog = () => {
  import("posthog-js").then((module) => {
    const posthog = module.default;

    if (!posthogInitialised) {
      posthog.init("<ph_project_api_key>", {
        api_host: "<ph_instance_address>",
        loaded: (posthog) => {
          if (process.env.NODE_ENV === "development") {
            posthog.opt_out_capturing();
          }
        },
      });
      posthogInitialised = true;
    }

    return module.default;
  });
};

function MyApp({ Component, pageProps }) {
  const router = useRouter();

  const handleRouteChange = async () => {
    (await asyncPosthog()).capture("$pageview");
  }

  useEffect(() => {
    router.events.on("routeChangeComplete", handleRouteChange);

    return () => {
      router.events.off("routeChangeComplete", handleRouteChange);
    };
  }, []);

  return <Component {...pageProps} />;
}

export default MyApp;

```

I would have to check if `handleRouteChange` is actually fired on initial load, otherwise this code doesn't work (posthog will not be loaded otherwise so there will be no pageview event).

**Comment by nbouvrette** [comment]
Sorry, @benjackwhite I meant misleading in terms of performance because adding 30kb to your bundle size could easily be avoided by amending the doc. Unless you are saying this can cause some sort of impact? We tested it on ur side and it seems fine so far, and our bundle size is also much better.

Also good catch for the example I provided - we actually have a wrapper on our side so I didn't have the opportunity to test it but we ended up having something closer to this (moving the import outside) which achieves similar results I would think:


```ts
import { useRouter } from 'next/router';
import { useEffect } from 'react';

const dynamicPostHogImport = import('posthog-js');

let posthogInitialised = false;

// Dynamically imports PostHog to track page views
const handleRouteChange = () => {
  void dynamicPostHogImport.then((module) => {
    if (!posthogInitialised) {
      const posthog = module.default
      posthog.init("<ph_project_api_key>", {
        api_host: "<ph_instance_address>",
        loaded: (posthog) => {
          if (process.env.NODE_ENV === "development") {
            posthog.opt_out_capturing();
          }
        },
      });
      posthogInitialised = true;
    }
  })
}

function MyApp({ Component, pageProps }) {
  const router = useRouter();

  useEffect(() => {
    router.events.on('routeChangeComplete', handleRouteChange);

    return () => {
      router.events.off('routeChangeComplete', handleRouteChange);
    };
  }, []);

  return <Component {...pageProps} />;
}

export default MyApp;
```

**Comment by benjackwhite** [comment]
Makes sense 👍  I'm hesitant to change the documentation as this to me still feels like a matter of tradeoffs and choosing which one you want:

For example if your bundle is normally 500kb:

**Option 1 - include in the bundle**
* Browser loads 530kb of JS (your code + posthog)
* `posthog` is always available

**Option 2 - don't include in the bundle**
* Browser loads 500kb of JS (your code)
* Browser loads another 30kb of JS (posthog)
* `posthog` is **not** always available and needs to be wrapped with dynamic import + async code

For most people Option 1 is totally acceptable and perhaps even preferable. Option 2 is an optimisation choice for the developer, probably in cases where the analytics being 100% reliable is less important than the time spent loading blocking JS.

I'm not sure that PostHog is unique here. The dynamic import could be applied to essentially any 3rd party code you would choose to implement and comes with the caveats that the framework in question would provide options to overcome

**Comment by nbouvrette** [comment]
Another idea, if you don't want to amend the Next.js doc directly could be to have an optimization doc? I saw a few of those for Google analytics where they explain how to use it in async mode.

Maybe you already have such a document but I couldn't find it the other day. It might also be useful to provide some examples in there for Next.js and other frameworks
